pkg/webhooks/scc/scc.go

```diff
@@ -0,0 +1,208 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	admissionctl "sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+const (
+	WebhookName string = "scc-validation"
+	docString   string = `Managed OpenShift Customers may not modify the following default SCCs: %s`
+)
+
+var (
+	timeout        int32 = 2
+	log                  = logf.Log.WithName(WebhookName)
+	anyuidPriority int32 = 10
+	scope                = admissionregv1.ClusterScope
+	rules                = []admissionregv1.RuleWithOperations{
+		{
+			Operations: []admissionregv1.OperationType{"CREATE", "UPDATE", "DELETE"},
+			Rule: admissionregv1.Rule{
+				APIGroups:   []string{"security.openshift.io"},
+				APIVersions: []string{"*"},
+				Resources:   []string{"securitycontextconstraints"},
+				Scope:       &scope,
+			},
+		},
+	}
+	defaultSCCs = []string{
+		"anyuid",
+		"hostaccess",
+		"hostmount-anyuid",
+		"hostnetwork",
+		"node-exporter",
+		"nonroot",
+		"privileged",
+		"restricted",
+		"pipelines-scc",
+	}
+)
+
+type SCCWebHook struct {
+	s runtime.Scheme
+}
+
+// NewWebhook creates the new webhook
+func NewWebhook() *SCCWebHook {
+	scheme := runtime.NewScheme()
+	admissionv1.AddToScheme(scheme)
+	corev1.AddToScheme(scheme)
+
+	return &SCCWebHook{
+		s: *scheme,
+	}
+}
+
+// Authorized implements Webhook interface
+func (s *SCCWebHook) Authorized(request admissionctl.Request) admissionctl.Response {
+	return s.authorized(request)
+}
+
+func (s *SCCWebHook) authorized(request admissionctl.Request) admissionctl.Response {
+	var ret admissionctl.Response
+
+	scc, err := s.renderSCC(request)
+	if err != nil {
+		log.Error(err, "Couldn't render a SCC from the incoming request")
+		return admissionctl.Errored(http.StatusBadRequest, err)
+	}
+
+	if request.Operation == admissionv1.Delete {
+		if isDefaultSCC(scc) {
+			ret = admissionctl.Denied("Deleting default SCCs is not allowed")
+			ret.UID = request.AdmissionRequest.UID
+			return ret
+		}
+	}
+
+	if request.Operation == admissionv1.Update {
+		if isDefaultSCC(scc) {
+			ret = admissionctl.Denied("Modifying default SCCs is not allowed")
+			ret.UID = request.AdmissionRequest.UID
+			return ret
+		}
+	}
+
+	if request.Operation == admissionv1.Create {
+		if SCCwithHigherPriority(scc) {
+			ret = admissionctl.Denied(fmt.Sprintf("Creating SCC with priority higher than %d is not allowed", anyuidPriority))
+			ret.UID = request.AdmissionRequest.UID
+			return ret
+		}
+	}
+
+	ret = admissionctl.Allowed("Request is allowed")
+	ret.UID = request.AdmissionRequest.UID
+	return ret
+}
+
+// renderSCC render the SCC object from the requests
+func (s *SCCWebHook) renderSCC(request admissionctl.Request) (*securityv1.SecurityContextConstraints, error) {
+	decoder, err := admissionctl.NewDecoder(&s.s)
+	if err != nil {
+		return nil, err
+	}
+	scc := &securityv1.SecurityContextConstraints{}
+	if len(request.OldObject.Raw) > 0 {
+		err = decoder.DecodeRaw(request.OldObject, scc)
```

Comment:
if a SCC is created with a priority of 1, and then a request is sent to update it to 20, will OldObject.Raw contain the 1 or 20?

Reply:
Yeah, that a good point, if we are not going to allow the creation of higher priority, we should also limit the updating as well.

Reply:
I see the latest commits address updates, but I'm still not sure if OldObject is the right choice for updating resources.

Reply:
Hi Manuel,
I update the `renderSCC` function to read both the old and new object.

So the validation will be:
```
when operation = Create:
    validate the newObject
when operation = Delete:
    validate the oldObject
when operation = Update:
    validate both newObject and oldObject
```

Reply:
cool. Would it be possible to cover this behavior with a test case? Did you verify the behavior on update? Is the old object the existing one and the new object the one sent to OpenShift?
Do we actually need to validate the old object if the new object is there?